File: src/tir/ir/event.cc
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

#include <tvm/ffi/reflection/registry.h>
#include <tvm/tir/event.h>

namespace tvm {
namespace tir {

TVM_FFI_STATIC_INIT_BLOCK({
  BulkGroupEventNode::RegisterReflection();
  SemaphoreEventTensorNode::RegisterReflection();
  SemaphoreEventTensorItemNode::RegisterReflection();
});

kEventImpl BulkGroupEventNode::GetImpl() const { return impl; }
ffi::Array<ffi::Any> BulkGroupEventNode::GetState() const { return state; }

BulkGroupEvent::BulkGroupEvent(kEventImpl impl, const ffi::Array<ffi::Any>& state,
                               const ffi::String& name) {
  ObjectPtr<BulkGroupEventNode> n = ffi::make_object<BulkGroupEventNode>();
  n->name = std::move(name);
  n->impl = impl;
  n->state = std::move(state);
  data_ = std::move(n);
}

SemaphoreEventTensor::SemaphoreEventTensor(const kEventImpl& impl,
                                           const ffi::Array<ffi::Any>& state,
                                           const ffi::Array<PrimExpr>& shape,
                                           const ffi::String& name) {
  ObjectPtr<SemaphoreEventTensorNode> n = ffi::make_object<SemaphoreEventTensorNode>();
  n->name = std::move(name);
  n->impl = impl;
  n->state = std::move(state);
  n->shape = std::move(shape);
  data_ = std::move(n);
}

kEventImpl SemaphoreEventTensorItemNode::GetImpl() const { return tensor->impl; }
ffi::Array<ffi::Any> SemaphoreEventTensorItemNode::GetState() const { return tensor->state; }

SemaphoreEventTensorItem::SemaphoreEventTensorItem(const SemaphoreEventTensor& tensor,
                                                   const ffi::Array<PrimExpr>& indices) {
  ObjectPtr<SemaphoreEventTensorItemNode> n = ffi::make_object<SemaphoreEventTensorItemNode>();
  n->tensor = std::move(tensor);
  n->indices = std::move(indices);
  data_ = std::move(n);
}

TVM_FFI_STATIC_INIT_BLOCK({
  namespace refl = tvm::ffi::reflection;
  refl::GlobalDef()
      .def("tirp.BulkGroupEvent",
           [](kEventImpl impl, ffi::Array<ffi::Any> state, ffi::String name) {
             return BulkGroupEvent(impl, state, name);
           })
      .def("tirp.SemaphoreEventTensor",
           [](kEventImpl impl, ffi::Array<ffi::Any> state, ffi::Array<PrimExpr> shape,
              ffi::String name) { return SemaphoreEventTensor(impl, state, shape, name); })
      .def("tirp.SemaphoreEventTensorItem",
           [](SemaphoreEventTensor tensor, ffi::Array<PrimExpr> indices) {
             return SemaphoreEventTensorItem(tensor, indices);
           })
      .def("tirp.BaseEventImplGet", [](BaseEvent event) { return event->GetImpl(); })
      .def("tirp.BaseEventStateGet", [](BaseEvent event) { return event->GetState(); });
});

}  // namespace tir
}  // namespace tvm

File: python/tvm/tir/event.py
# Licensed to the Apache Software Foundation (ASF) under one
# or more contributor license agreements.  See the NOTICE file
# distributed with this work for additional information
# regarding copyright ownership.  The ASF licenses this file
# to you under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in compliance
# with the License.  You may obtain a copy of the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing,
# software distributed under the License is distributed on an
# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
# KIND, either express or implied.  See the License for the
# specific language governing permissions and limitations
# under the License.
"""TIRp event"""

from enum import IntEnum
from typing import List, Any, Dict, Optional

import tvm_ffi

from tvm.runtime import Object
from ..tirp import _ffi_api
from tvm.tir.stmt import OpCall
from tvm.ir import Op
from tvm.tir import PrimExpr


def make_op_call(
    op_name: str,
    *args,
    dispatch: Optional[str] = None,
    config: Dict[str, Any] = None,
):
    assert isinstance(op_name, str)
    f = tvm_ffi.get_global_func("script.ir_builder.tir.OpCall")
    return f(OpCall(*args, op=Op.get("tirp." + op_name), config=config, dispatch=dispatch))


class EventImpl(IntEnum):
    # see also include/tvm/tir/event.h kEventImpl
    kCpAsync = 0
    kTMALoad = 1  # state: mbarrier, phase, tx_cnt
    kTMAStore = 2
    kGlobalSemaphore = 3  # state sem, state


@tvm_ffi.register_object("tirp.BaseEvent")
class BaseEvent(Object):
    """Base event"""

    def __init__(self, name: str):
        raise NotImplementedError("BaseEvent is not instantiable")

    def get_impl(self) -> EventImpl:
        return _ffi_api.BaseEventImplGet(self)

    def get_state(self) -> List[Any]:
        return _ffi_api.BaseEventStateGet(self)


@tvm_ffi.register_object("tirp.BulkGroupEvent")
class BulkGroupEvent(BaseEvent):
    """Bulk group event"""

    impl: EventImpl
    state: List[Any]
    name: str

    def __init__(self, impl: EventImpl, state: List[Any] = [], name: str = ""):
        self.__init_handle_by_constructor__(_ffi_api.BulkGroupEvent, impl, state, name)

    def init(self):
        return make_op_call("event_init", self)

    def commit(self, dispatch: Optional[str] = None, **kwargs):
        return make_op_call("event_commit", self, dispatch=dispatch, config=kwargs)

    def wait(self, n_groups: int = 0, dispatch: Optional[str] = None, **kwargs):
        return make_op_call("event_wait", self, n_groups, dispatch=dispatch, config=kwargs)


@tvm_ffi.register_object("tirp.SemaphoreEventTensorItem")
class SemaphoreEventTensorItem(BaseEvent):
    """Event tensor item"""

    tensor: "SemaphoreEventTensor"
    indices: List[PrimExpr]

    def __init__(self, tensor: "SemaphoreEventTensor", indices: List[PrimExpr]):
        self.__init_handle_by_constructor__(_ffi_api.SemaphoreEventTensorItem, tensor, indices)

    def init(self, expected_count: PrimExpr, dispatch: Optional[str] = None, **kwargs):
        return make_op_call("event_init", self, expected_count, dispatch=dispatch, config=kwargs)

    def commit(self, dispatch: Optional[str] = None, **kwargs):
        return make_op_call("event_commit", self, dispatch=dispatch, config=kwargs)

    def wait(self, dispatch: Optional[str] = None, **kwargs):
        return make_op_call("event_wait", self, dispatch=dispatch, config=kwargs)


@tvm_ffi.register_object("tirp.SemaphoreEventTensor")
class SemaphoreEventTensor(Object):
    """Event tensor"""

    impl: EventImpl
    state: List[Any]
    name: str
    shape: List[PrimExpr]

    def __init__(
        self, impl: EventImpl, state: List[Any] = [], shape: List[PrimExpr] = [], name: str = ""
    ):
        self.__init_handle_by_constructor__(_ffi_api.SemaphoreEventTensor, impl, state, shape, name)

    def __getitem__(self, *indices: List[PrimExpr]):
        assert len(indices) == len(
            self.shape
        ), "indices must be the same length as shape, but got {} and {}".format(
            len(indices), len(self.shape)
        )
        return SemaphoreEventTensorItem(self, indices)

    def init(self, expected_count: PrimExpr, dispatch: Optional[str] = None, **kwargs):
        return make_op_call("event_init", self, expected_count, dispatch=dispatch, config=kwargs)

    def get_impl(self) -> EventImpl:
        return self.impl

    def get_state(self) -> List[Any]:
        return self.state
